pkg/apis/core/v1beta1/validation/validation.go

```diff
@@ -144,6 +149,88 @@ func ValidateFederatedTypeConfigStatus(status *v1beta1.FederatedTypeConfigStatus
 	return allErrs
 }
 
+func ValidateKubeFedCluster(obj *v1beta1.KubeFedCluster, statusSubResource bool) field.ErrorList {
+	var allErrs field.ErrorList
+	if !statusSubResource {
+		allErrs = validateKubeFedClusterSpec(&obj.Spec, field.NewPath("spec"))
+	} else {
+		allErrs = validateKubeFedClusterStatus(&obj.Status, field.NewPath("status"))
+	}
+	return allErrs
+}
+
+func validateKubeFedClusterSpec(spec *v1beta1.KubeFedClusterSpec, fldPath *field.Path) field.ErrorList {
+	allErrs := validateAPIEndpoint(spec.APIEndpoint, fldPath.Child("apiEndpoint"))
+	allErrs = append(allErrs, validateLocalSecretReference(&spec.SecretRef, fldPath.Child("secretRef"))...)
+	return allErrs
+}
+
+func validateKubeFedClusterStatus(status *v1beta1.KubeFedClusterStatus, fldPath *field.Path) field.ErrorList {
+	var allErrs field.ErrorList
+
+	for i, condition := range status.Conditions {
+		allErrs = append(allErrs, validateClusterCondition(&condition, fldPath.Child("conditions").Index(i))...)
+	}
+	return allErrs
+}
+
+func validateAPIEndpoint(endpt string, fldPath *field.Path) field.ErrorList {
+	if endpt == "" {
+		return field.ErrorList{field.Required(fldPath, "")}
+	}
+
+	// Parse APIEndpoint using the same mechanism used by client-go
+	// when the controller-manager creates a client using this APIEndpoint
+	// value.
+	hostURL, _, err := rest.DefaultServerURL(endpt, "", schema.GroupVersion{}, true)
+	if err != nil {
+		return field.ErrorList{field.Invalid(fldPath, endpt, err.Error())}
+	}
+
+	allErrs := validateEnumStrings(fldPath, hostURL.Scheme, []string{"http", "https"})
+
+	hostname := hostURL.Hostname()
+	dnsErrs := valutil.IsDNS1123Subdomain(hostname)
+	ipErrs := valutil.IsValidIP(hostname)
+	if dnsErrs != nil && ipErrs != nil {
```

Comment:
I can deduce from this logic that the hostname needs to validate either as a valid subdomain or ip, and that a validation error for both paths indicates an invalid hostname.  I'm not sure both sets of errors will make much sense to the user in the form presented, though.  The test checks for one or the other, but the user would be presented by both error conditions.  What do you think of adding a generic error message instead, similar to what `DefaultServerURL()` returns?

Reply:
I think that's a good compromise!